Assets/Scripts/PlayerController.cs

```diff
@@ -1,39 +1,92 @@
 ﻿using System.Collections;
 using System.Collections.Generic;
-using System.Diagnostics;
 using System.Threading;
+using Unity.VisualScripting;
 using UnityEngine;
+using UnityEngine.UIElements;
 
-[RequireComponent(typeof(Rigidbody2D))]
+[RequireComponent(typeof(Rigidbody2D), typeof(Collider2D))]
 public class PlayerCharacterScript : MonoBehaviour
 {
     [Tooltip("Speed in units per second")]
     public float Speed;
 
-    /// <summary>
-    /// Rigidbody without gravity
-    /// </summary>
     private Rigidbody2D _rigidbody2D;
+    private Collider2D _collider;
+    private float _horizontalInput = 0f;
+    private float _verticalInput = 0f;
 
     // Start is called before the first frame update
     void Start()
     {
-        this._rigidbody2D = GetComponent<Rigidbody2D>();
+        _rigidbody2D = GetComponent<Rigidbody2D>();
+        _collider = GetComponent<Collider2D>();
     }
 
     // FixedUpdate is called once per physics frame
     void FixedUpdate()
     {
-        float horizontalInput = Input.GetAxisRaw("Horizontal");
-        float verticalInput = Input.GetAxisRaw("Vertical");
+        _horizontalInput = Input.GetAxisRaw("Horizontal");
+        _verticalInput = Input.GetAxisRaw("Vertical");
 
-        Vector2 horizontalInputVector = new Vector2(horizontalInput, 0).normalized;
-        Vector2 verticalInputVector = new Vector2(0, verticalInput).normalized;
-        Vector2 direction = (horizontalInputVector + verticalInputVector).normalized;
+        float halfColliderWidth = _collider.bounds.extents.x;
```

Comment:
Do you really need to code the collisions yourself? Unity's Ridigbody can be toggled to block everything

Reply:
One reason that I didn't want to leave it to rigidbody collisions is this usecase:
The speed is set to 10 units per second, so if the player moves along a single axis, they'll move 10 units per second on that axis.
If they move diagonally, then they'll move 10 units per second diagonally, or about 7.07 units per second per axis (due to trigonometry)
However, with using just a rigidbody border, it makes the player seem to "drag" along the edges, the player moves slower than expected.
For example, if you press W and A while in the lower-left corner, you'll only move vertically 7.07 units per second rather than the desired 10 units per second.
My code handles that correctly, where a rigidbody border wouldn't.